### PR TITLE
1567914: String truncation shouldn't truncate in the middle of a codepoint

### PR DESCRIPTION
**Pull request by mdboom**
This works by starting one character beyond the truncation point and moves back until it finds a character boundary.  The string is then truncated to the byte immediately preceding that boundary.

An alternative would have been to iterate over characters from the beginning, but that would be O(n), whereas this is O(1) (since barring corrupted string data, it should only have to look at at most 3 bytes).

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/src/metrics/string.rs [hunk 1]
```diff
@@ -51,18 +51,7 @@ impl StringMetric {
...
-                "Value length {} exceeds maximum of {}",
-                s.len(),
-                MAX_LENGTH_VALUE
-            );
-            record_error(glean, &self.meta, ErrorType::InvalidValue, msg);
-            s[0..MAX_LENGTH_VALUE].to_string()
-        } else {
-            s
-        };
+        let s = truncate_string_at_boundary_with_error(glean, &self.meta, value, MAX_LENGTH_VALUE);
```
Comment (Dexterp37) [inline glean-core/src/metrics/string.rs]: Would you kindly also update [experiments](https://github.com/mozilla/glean/blob/104d49e8d9fd215e02469c09f6c0a8ab4e63f5e6/glean-core/src/metrics/experiment.rs#L60)?

Comment (mdboom) [reply glean-core/src/metrics/string.rs]: Thanks.  Missed that one.

**Comment by mdboom** [comment]
Thanks.  This is Rusty enough that I'll wait for an approval from @badboy as well.

File: glean-core/src/util.rs [hunk 2]
```diff
@@ -121,4 +183,19 @@ mod test {
             fixed_now
         );
     }
+
+    #[test]
```
Comment (badboy) [inline glean-core/src/util.rs]: Can we get one test on an ASCII value, just to make sure? e.g. `"1234567890abcdef"` truncated to 10.

**Comment by badboy** [comment]
(actually approving, r+wc)

**Pull request merged**